orttraining/orttraining/core/framework/module_gradient_graph_builder.cc

```diff
@@ -0,0 +1,393 @@
+    split_graphs_info_.user_output_grad_names.emplace_back(output_gradient_name);
+
+    // Only add to graph input when it's not an output of a node.
+    if (output_names.find(output_gradient_name) == output_names.end()) {
+      split_graphs_info_.backward_output_grad_names.emplace_back(output_gradient_name);
+      NodeArg* output_gradient_node_arg = graph.GetNodeArg(output_gradient_name);
+      output_gradient_node_arg->UpdateTypeAndShape(*graph.GetNodeArg(output_name), true, true, *logger_);
+      input_args.emplace_back(output_gradient_node_arg);
+    }
+  }
+
+  graph.SetInputs(input_args);
+
+  std::vector<const NodeArg*> output_args;
+  for (auto& output_name : split_graphs_info_.user_output_names) {
+    output_args.emplace_back(graph.GetNodeArg(output_name));
+  }
+
+  // Add input gradients to graph outputs if it's required.
+  split_graphs_info_.user_input_grad_names.clear();
+  for (const auto& input_name : config_.input_names_require_grad) {
+    std::string input_gradient_name = input_name + "_grad";
+    ORT_ENFORCE(output_names.find(input_gradient_name) != output_names.end(),
+                "Required user input grad is not found on gradient graph.");
+    split_graphs_info_.user_input_grad_names[input_name] = input_gradient_name;
+    output_args.emplace_back(graph.GetNodeArg(input_gradient_name));
+  }
+
+  // Add initializer gradients to graph outputs.
+  split_graphs_info_.initializer_grad_names_to_train.clear();
+  for (const auto& initializer_name : split_graphs_info_.initializer_names_to_train) {
+    std::string initializer_gradient_name = initializer_name + "_grad";
+    ORT_ENFORCE(output_names.find(initializer_gradient_name) != output_names.end(),
+                "Trainable initializer grad is not found on gradient graph.");
+    split_graphs_info_.initializer_grad_names_to_train.emplace_back(initializer_gradient_name);
+    output_args.emplace_back(graph.GetNodeArg(initializer_gradient_name));
+  }
+
+  graph.SetOutputs(output_args);
+  ORT_RETURN_IF_ERROR(graph.Resolve());
+
+  // Run the transformers again mainly for backward part, e.g., constant fold from those Shape nodes in backward graph.
+  // TODO: maybe we need another list of transformers here rather than the list for pre-training list.
+  for (int i = static_cast<int>(TransformerLevel::Level1); i <= static_cast<int>(TransformerLevel::MaxLevel); i++) {
+    ORT_RETURN_IF_ERROR(graph_transformation_mgr.ApplyTransformers(graph, static_cast<TransformerLevel>(i), *logger_));
+  }
+
+  // Create two copies of gradient model for forward and backward models respectively.
+  auto gradient_model_proto = model_copied->ToProto();
+  ORT_RETURN_IF_ERROR(Model::Load(gradient_model_proto, forward_model_, nullptr, *logger_));
+  ORT_RETURN_IF_ERROR(Model::Load(gradient_model_proto, backward_model_, nullptr, *logger_));
+
+  // Split the graph in the copies of gradient model.
+  ORT_RETURN_IF_ERROR(Split());
```

Comment:
I feel the overall workflow can be simplified quite a bit, but this is a non-blocking optimization item 

In GradientGraphBuilder::Build(), we first construct a gradient_graph_defs, that is effectively a "description" of bw graph. If we perform the final GraphAugmenter::AugmentGraph() step directly on an empty model, the resulting graph is essentially the bw graph. So there will no need for Splitting step. 

Of course, we still need to populate the graph inputs/outputs properly on fw and bw graph, but I think both pieces of information are fully captured in the gradient_graph_defs.